### DISCUSSION TITLE
아이템 63. 문자열 연결은 느리니 주의하라

### DISCUSSION THREAD

**Discussion by coalong** [Study-2-Effective-Java/Effective-Java#146] [3. 과제] [answered]
[Java | String 연산.pdf](https://github.com/Study-2-Effective-Java/Effective-Java/files/10966959/Java.String.pdf)

## TL; DR
> 문자열을 연결할 때는 문자열 연결 연산자(+) 대신 StringBuilder 의 append 메서드나 문자 배열을 사용하라.

<br/>

## 문자열 연결 연산자(+) 
- 문자열 연결 연산자(+)로 문자열 n개를 잇는 시간은 n<sup>2</sup>에 비례한다.
- **_문자열은 불변_(아이템 17)** 이라서 두 문자열을 연결할 경우 양쪽의 내용을 모두 복사해야 하므로 성능 저하가 일어난다.

```java
String str = "a";

for (int i = 0; i < 100_000; i++) {
    str += "a";
}
```

<br/>

## StringBuilder 를 통한 문자열 조작
- 성능을 포기하고 싶지 않다면 String 대신 StringBuilder 를 사용하자.

```java
StringBuilder builder = new StringBuilder("a");

for (int i = 0; i < 100_000; i++) {
    builder.append("a");
}
```

```bash
// 10만번 실행(위)
String 연산자(+) 실행시간: 1339
StringBuilder 실행시간: 3

// 50만번 실행(아래)
String 연산자(+) 실행시간: 27293
StringBuilder 실행시간: 14
```
<aside>
💡 String 연산자(+) 는 제곱이 비례해 늘어나고 StringBuilder 는 선형으로 늘어난다.
</aside>

<br/>
<br/>
<br/>

## StringBuilder VS StringBuffer
```java
// StringBuffer 👉 synchronized 키워드가 있다.
@Override
public synchronized StringBuffer append(Object obj) {
    toStringCache = null;
    super.append(String.valueOf(obj));
    return this;
}

// StringBuilder 👉 synchronized 키워드가 없다.
@Override
public StringBuilder append(Object obj) {
    return append(String.valueOf(obj));
}
```
![Untitled 2](https://user-images.githubusercontent.com/71637466/224968756-da24e3cd-fab5-47e7-a402-aed32ecd7278.png)

<br/>

## 문자열 연산 속도 비교
![Untitled 3](https://user-images.githubusercontent.com/71637466/224968847-c1300841-a8a8-4053-8cf3-aabac4c44f9e.png)

**하지만 대부분의 상황에서 `StringBuilder`를 사용할 때** 
**단일 쓰레드만 이용하기 때문에 실제 `StringBuffer`를 사용할 일은 거의 없다.**
> *오히려 싱글쓰레드 상황에서 StringBuffer를 사용하게 되면 불필요한 동기화 과정을 거치기 때문에 성능만 더 안좋아질 뿐이에요.*
>   

<br/>
<br/>

### ***왜 IntelliJ 에서 StringBuilder 보다 String 을 추천하는거지?***
<img width="562" alt="Untitled 4" src="https://user-images.githubusercontent.com/71637466/224969163-77378ce5-20d9-419a-8e95-67b2d815f13b.png">

### ***이유는?***

> j**ava 1.5 이후부터는 컴파일 시점**에 (+)연산자의 문자열 연결을 StringBuilder 인스턴스를 생성하여 같은 동작을 한다.
→ 같은 동작을 하여 성능상 이점이 없기 때문에 가독성을 고려한 String 을 추천한다.
> 

   

### 단, 동적 문자열 연결은 (ex. loop문) StringBuilder 를 직접 생성해야 한다.

```java
StringBuilder result = new StringBuilder("data");

for (int i = 0; i < 100; i++) {
    result.append("some more data");
}
```

<br/>
<br/>

## 참고자료
[유용한 클래스들](https://corock.tistory.com/288)
[간단한 문자열 연결](https://dzone.com/articles/string-concatenation-performacne-improvement-in-ja)
[java 8을 사용한 문자열 연결](https://pellegrino.link/2015/08/22/string-concatenation-with-java-8.html)

**Comment by corock** [comment, marked as answer]
질문하려고 한 부분(`StringBuilder` 보다 `String` 을 추천하는 이유)까지 정리하셨네요.
`String` / `StringBuffer` / `StringBuilder` 클래스 간 비교나 장단점은 기술 면접에서도 자주 물어보는 주제 중 하나로 알고 있습니다.

내용 잘 읽었습니다~ ㅎㅎ

**Reply by JoisFe** [reply]
오호 새롭게 알았습니다  String의 + 연산자로의 문자열 연결이 StringBuilder 인스턴스를 생성해 같은 동작을 한다니...😀
